sql/open_address_hash.h

```diff
@@ -0,0 +1,316 @@
+#pragma once
+
+#include <string.h>
+#include <my_dbug.h>
+
+
+
+template <typename key_trait, typename trait> class open_address_hash
+{
+public:
+  using T= typename trait::elem_type;
+  using find_type= typename trait::find_type;
+  using erase_type= typename trait::erase_type;
+  using hash_value_type= typename key_trait::hash_value_type;
+  using key_type= typename key_trait::key_type;
+
+  //key_type *get_key(const T &elem) { return key_trait::get_key(elem); }
+  bool is_empty(const T &el) { return trait::is_empty(el); }
+  void set_null(T &el) { trait::set_null(el); }
+
+  open_address_hash()
+  {
+    first.set_mark(true);
+    first.set_ptr(nullptr);
+    second= nullptr;
+  }     
+
+  ~open_address_hash()
+  {
+    if (!first.mark())
+    {
+      DBUG_ASSERT(hash_array);
+      free(hash_array);
+    }
+  }
+
+private:
+  hash_value_type to_index(const hash_value_type &hash_value)
+  {
+    return hash_value & (capacity - 1);
+  }
+
+  hash_value_type hash_from_value(const T &value)
+  {
+    return trait::get_hash_value(value);
+  }
+
+  bool insert_helper(const T &value)
+  {
+
+    auto hash_val= to_index(hash_from_value(value));
+
+    while (hash_array[hash_val] != nullptr)
+    {
+      if (hash_array[hash_val] == value)
+        return false;
+      hash_val= to_index(hash_val + 1);
+    }
+
+    hash_array[hash_val]= value;
+    _size++;
+    return true;
+  };
+
+  uint rehash_subsequence(uint i)
+  {
+    for (uint j= to_index(i + 1); hash_array[j] != nullptr; j= to_index(j + 1))
+    {
+      auto temp_el= hash_array[j];
+      hash_array[j]= nullptr;
+      //_size--;
+      insert_helper(temp_el);
+      /*auto key= to_index(key_trait::get_hash_value(trait::get_key(hash_array[j])));
+      if (key <= i || key > j)
+      {
+        hash_array[i]= hash_array[j];
+        i= j;
+      }*/
+    }
+
+    return i;
+  }
+
+  bool erase_helper(const erase_type &value)
+  {
+    for (auto key= to_index(trait::get_hash_value(value));
+         hash_array[key] != nullptr; key= to_index(key + 1))
+    {
+      if (trait::is_equal(hash_array[key], value))
+      {
+        hash_array[key]= nullptr;
+        _size--;
+        rehash_subsequence(key);
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  void rehash(const uint _capacity)
+  {
+    uint past_capacity= capacity;
+    capacity= _capacity;
+
+    if (past_capacity > capacity)
+    {
+      _size= 0;
```

Comment:
should we tear this _size value back ad forth during rehash?
It seems that the only place it should be increased is `insert`!

Same for the decrease -- just put is in `remove` and clean these torturings elsewere.